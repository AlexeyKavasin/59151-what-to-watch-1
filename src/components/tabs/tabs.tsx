import * as React from "react";
import {Tab} from "../tab/tab";
import {connect} from "react-redux";
import {TabContent} from "../tab-content/tab-content";
import {ITabs, ITabsState} from "../../interfaces";

class Tabs extends React.Component<ITabs, ITabsState> {
    constructor(props) {
        super(props);
        this.state = {
            activeTab: 2
        }

        this.onTabClick = this.onTabClick.bind(this);
        this.formatDate = this.formatDate.bind(this);
    }
    
    onTabClick(evt: React.MouseEvent<HTMLElement>, tabId: number): void {
        evt.preventDefault();
        if (tabId && tabId !== this.state.activeTab) {
            this.setState({
                activeTab: tabId
            })
        }
    }

    formatDate(date: string): string {
      const d = new Date(date).getDate();
      const m = new Date(date).getMonth() + 1;
      const y = new Date(date).getFullYear();

      return [d, m, y].map((val: number) => val > 10 ? val : `0${val}`).join('.');
    }

    getRatingText(rating: number): string {
        switch(true) {
            case(rating === 10):
                return `Awesome`;
            case(rating >= 8 && rating < 10):
                return `Very good`;
            case(rating >= 5 && rating < 8):
                return `Good`;
            case(rating >= 3 && rating < 5):
                return `Normal`;
            case(rating >= 0 && rating < 3):
                return `Bad`;
            default:
                return `Sorry no rating`;
        }
    }

    render() {
        const {film, comments} = this.props;

        return (
          <React.Fragment>
            <nav className="movie-nav movie-card__nav">
              <ul className="movie-nav__list">
                <Tab
                  tabId={1}
                  activeTab={this.state.activeTab}
                  onTabClick={this.onTabClick}
                >
                  Overview
                </Tab>
                <Tab
                  tabId={2}
                  activeTab={this.state.activeTab}
                  onTabClick={this.onTabClick}
                >
                  Details
                </Tab>
                <Tab
                  tabId={3}
                  activeTab={this.state.activeTab}
                  onTabClick={this.onTabClick}
                >
                  Reviews
                </Tab>
              </ul>
            </nav>

            <TabContent tabId={1} activeTab={this.state.activeTab}>
              <div className="movie-rating">
                <div className="movie-rating__score">{film.rating}</div>
                <p className="movie-rating__meta">
                  <span className="movie-rating__level">{this.getRatingText(film.rating)}</span>
                  <span className="movie-rating__count">
                    {film.scores_count} ratings
                  </span>
                </p>
              </div>
              <div className="movie-card__text">
                <p>{film.description}</p>
                <p className="movie-card__director">
                  <strong>Director: {film.director}</strong>
                </p>
                <p className="movie-card__starring">
                  <strong>
                    Starring:
                    {film.starring
                      .map((actor: string) => actor)
                      .join(`, `)}
                  </strong>
                </p>
              </div>
            </TabContent>

            <TabContent tabId={2} activeTab={this.state.activeTab}>
              <div className="movie-card__text movie-card__row">
                <div className="movie-card__text-col">
                  <p className="movie-card__details-item">
                    <strong className="movie-card__details-name">
                      Director
                    </strong>
                    <span className="movie-card__details-value">
                      {film.director}
                    </span>
                  </p>
                  <p className="movie-card__details-item">
                    <strong className="movie-card__details-name">
                      Starring
                    </strong>
                    <span className="movie-card__details-value">
                      {film.starring
                        .map((actor: string) => actor)
                        .join(`, `)}
                    </span>
                  </p>
                </div>
                <div className="movie-card__text-col">
                  <p className="movie-card__details-item">
                    <strong className="movie-card__details-name">
                      Run Time
                    </strong>
                    <span className="movie-card__details-value">
                      {film.run_time}
                    </span>
                  </p>
                  <p className="movie-card__details-item">
                    <strong className="movie-card__details-name">
                      Genre
                    </strong>
                    <span className="movie-card__details-value">
                      {film.genre}
                    </span>
                  </p>
                  <p className="movie-card__details-item">
                    <strong className="movie-card__details-name">
                      Released
                    </strong>
                    <span className="movie-card__details-value">
                      {film.released}
                    </span>
                  </p>
                </div>
              </div>
            </TabContent>

            <TabContent tabId={3} activeTab={this.state.activeTab}>
              <div className="movie-card__reviews movie-card__row">
                <div className="movie-card__reviews-col">
                {comments && comments.map((comment, index) => {
                  return index % 2 === 0 ? <React.Fragment key={`comment-${index}`}>
                    <div className="review">
                      <blockquote className="review__quote">
                        <p className="review__text">
                          {comment.comment}
                        </p>

                        <footer className="review__details">
                          <cite className="review__author">
                            {comment.user.name}
                          </cite>
                          <time
                            className="review__date"
                            dateTime={comment.date}
                          >
                            {this.formatDate(comment.date)}
                          </time>
                        </footer>
                      </blockquote>

                      <div className="review__rating">{comment.rating}</div>
                    </div>
                  </React.Fragment> : null;
                })}
                </div>
                <div className="movie-card__reviews-col">
                {comments && comments.map((comment, index) => {
                  return index % 2 !== 0 ? <React.Fragment key={`comment-${index}`}>
                    <div className="review">
                      <blockquote className="review__quote">
                        <p className="review__text">
                          {comment.comment}
                        </p>

                        <footer className="review__details">
                          <cite className="review__author">
                            {comment.user.name}
                          </cite>
                          <time
                            className="review__date"
                            dateTime={comment.date}
                          >
                            {this.formatDate(comment.date)}
                          </time>
                        </footer>
                      </blockquote>

                      <div className="review__rating">{comment.rating}</div>
                    </div>
                  </React.Fragment> : null;
                })}
                </div>
              </div>
            </TabContent>
          </React.Fragment>
        );
    }
}

const mapStateToProps = (state, ownProps) => ({
  comments: state[`DATA`].comments
});

export {Tabs};

export default connect(
    mapStateToProps,
    null
)(Tabs);
