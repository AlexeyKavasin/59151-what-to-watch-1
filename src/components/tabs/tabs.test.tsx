import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Tabs} from './tabs';

const film = {
    name: `Matrix`,
    poster: `https://es31-server.appspot.com/wtw/static/film/preview/matrix.jpg`,
    id: 0,
    genre: `Action`,
    preview_image: `https://es31-server.appspot.com/wtw/static/film/preview/matrix.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    released: 1999,
    background_image: `https://es31-server.appspot.com/wtw/static/film/preview/matrix.jpg`,
    starring: [`Actor1`, `Actor2`],
    run_time: 120,
    director: `Director`,
    rating: 10,
    scores_count: 1000,
    description: `Lorem ipsum dolor sit amet, consectetur adipisicing elit. Numquam, soluta dolor? Cumque perferendis quaerat natus sapiente accusantium sequi. Ullam itaque, voluptatem nostrum quae fugiat voluptates necessitatibus aperiam eum numquam officiis.`
};

describe(`Tabs snapshot tests`, () => {
  it(`Tabs render correctly`, () => {
    const tree = renderer
      .create(
      <Tabs film={film}/>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
