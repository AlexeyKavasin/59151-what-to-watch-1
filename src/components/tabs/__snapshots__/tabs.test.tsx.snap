// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs snapshot tests Tabs render correctly 1`] = `
Array [
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>,
  <div
    className="movie-card__text movie-card__row"
  >
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Director
        </strong>
        <span
          className="movie-card__details-value"
        >
          Director
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Starring
        </strong>
        <span
          className="movie-card__details-value"
        >
          Actor1, Actor2
        </span>
      </p>
    </div>
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Run Time
        </strong>
        <span
          className="movie-card__details-value"
        >
          120
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Genre
        </strong>
        <span
          className="movie-card__details-value"
        >
          Action
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Released
        </strong>
        <span
          className="movie-card__details-value"
        >
          1999
        </span>
      </p>
    </div>
  </div>,
]
`;
